### PR TITLE
Introduce `ack-generate webhooks` sub command

### PR DESCRIPTION
**Pull request by A-Hilaly**
Part of https://github.com/aws-controllers-k8s/community/issues/835

This patch brings in a new sub command that can be used to generate
webhook functions. For now it can only generate conversion webhook
functions, but it can be extended to generate defaulting and validating
functions.

This patch also introduces a new function that generates conversion
functions. It can be used to generate `ConvertTo` and `ConvertFrom`
functions.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by A-Hilaly** [comment]
NOTE to reviews: This will need a rebase after #120 and #121 merge.
/hold

**Comment by RedbackThomson** [comment]
Could we hold off on this command until after I propose the metadata configuration? Options such as deprecated versions, etc. will all be available in that configuration file and it will not be necessary to pass them as command line arguments.

**Comment by A-Hilaly** [comment]
> Could we hold off on this command until after I propose the metadata configuration? Options such as deprecated versions, etc. will all be available in that configuration file and it will not be necessary to pass them as command line arguments.

Sure. I will rebase on top of your work once it's available.

File: pkg/generate/ack/output.go
Comment (A-Hilaly) [inline pkg/generate/ack/output.go:164]: @RedbackThomson If you can implement a similar function in your PR, i'm taking it :)

**Comment by A-Hilaly** [comment]
Need to be rebased, now that #126 is merged
/hold

File: pkg/generate/code/converter/converter.go
Comment (jaypipes) [inline pkg/generate/code/converter/converter.go:137]: This code doesn't look correct to me... Shouldn't it be like this?

```go
elementCopy := &v1alpha2.WebhookFilter{}
if element != nil {
    elementCopy.ExcludeMatchedPattern = element.ExcludeMatchedPattern
    elementCopy.Pattern = element.Pattern
    elementCopy.Type = element.Type
}
```

File: pkg/generate/code/converter/converter.go
Comment (jaypipes) [inline pkg/generate/code/converter/converter.go:319]: Why not just use `for _, %s := range %s {\n` here and get rid of the next line.

File: pkg/generate/code/converter/converter.go
Comment (jaypipes) [inline pkg/generate/code/converter/converter.go:384]: s/copyMap/copyList`

File: pkg/generate/code/converter/converter.go
Comment (jaypipes) [inline pkg/generate/code/converter/converter.go:336]: You are going to run into problems with nested list types here, because you will have variable name shadowing in effect. This is why the code generation that sets CR fields from an Output shape (the `pkg/generate/code.SetOutput` stuff) generates a variable name that represents the field's numeric index within the Shape. For instance, if I had a Shape like this:

```go
type Bar struct {
     Baz string
}

type Foo struct {
  Bars []*Bar
}

type MyShape struct {
  Foos []*Foo
}
```

The variable referring to `MyShape.Foos.Bars` would be called `f0f0` and the iterator over elements in that list of `Bar` structs would be called `f0f0iter`, with the element variable name of `f0f0elem`.

The way you are naming things here means that you would end up producing Go code like this:

```go
elementCopy := make([]*Foo, 0, len(src.Foos))
for _, element := range src.Foos {
    fooCopy := &Foo{}
    elementCopy := make([]*Bar, 0, len(element.Bars)
    for _, element := range element.Bars {
        barCopy := &Bar{}
        barCopy.Baz = element.Baz
        elementCopy = append(elementCopy, barCopy)
    }
    elementCopy = append(elementCopy, fooCopy)
}
```

which would be invalid since the variable name `elementCopy` is shadowed/hidden by the inner declaration of itself.

File: pkg/generate/code/converter/converter.go
Comment (jaypipes) [inline pkg/generate/code/converter/converter.go:474]: same comment as above :) you are going to run into variable naming conflicts.

File: pkg/generate/code/converter/converter.go
Comment (jaypipes) [inline pkg/generate/code/converter/converter.go:676]: s/retruns go/returns Go

**Review by jaypipes — CHANGES_REQUESTED** [review]
@A-Hilaly wonderful start on this! You will need to make a number of changes, however, in order to properly deal with nested complex types. See inline for an explanation of why your variable-naming schema needs to be adapted.

File: cmd/ack-generate/command/webhook.go
Comment (RedbackThomson) [reply cmd/ack-generate/command/webhook.go:76]: Even better, this can come directly from the `metadata` object.

File: pkg/generate/ack/output.go
Comment (RedbackThomson) [reply pkg/generate/ack/output.go:164]: These have been moved to `pkg/metadata`

**Pull request closed without merging**
